client/state/initial-state.js

```diff
@@ -157,16 +180,19 @@ function isValidReduxKeyAndState( key, state ) {
 	// able to force the state in memory to be rebuilt - possibly using
 	// https://stackoverflow.com/questions/35622588/how-to-reset-the-state-of-a-redux-store/35641992#35641992
 	// - without generating any errors. Until then, it must remain in place.)
-	const userId = get( state, [ 'currentUser', 'id' ], null );
+	const userId = state?.currentUser?.id ?? null;
 	return key === getReduxStateKeyForUserId( userId );
 }
 
-function persistentStoreState( reduxStateKey, storageKey, state, _timestamp ) {
+async function persistentStoreState( reduxStateKey, storageKey, state, _timestamp ) {
 	if ( storageKey !== 'root' ) {
 		reduxStateKey += ':' + storageKey;
 	}
 
-	return setStoredItem( reduxStateKey, Object.assign( {}, state, { _timestamp } ) );
+	const newState = { ...state, _timestamp };
```

Comment:
should we worry about overlapped writes here? Imagine we call 
```
persistenStoreState( 'foo', 'foo', ..., 1 );
persistenStoreState( 'foo', 'foo', ..., 2 );
```

Can the first await return after the second await? What impact will it have?

Reply:
I'm not sure the onus is on the async function to handle that situation. If you're invoking an async function twice in parallel, you should be prepared for execution to happen in parallel, in any order. The correct way to invoke it twice would be:

```js
await persistentStoreState( 'foo', 'foo', ..., 1 );
await persistentStoreState( 'foo', 'foo', ..., 2 );
```

Reply:
Right, but imagine two different processes are invoking it in a short time window...

Reply:
My understanding is that the only thing that calls `persistentStoreState` is `persistOnChange`, and it does so with different keys, therefore causing no collisions. `persistOnChange` itself is only called once, on boot.

If we still think that there could be issues (e.g. if I'm misreading `persistOnChange` and there could indeed be key collisions), then `persistOnChange` is the right place to address them, by awaiting each write in sequence.

Reply:
Yeah, I'm probably overthinking this. The big case that's turning around in my head is multiple tabs writing at the same time, but even that isn't really an issue since we don't read out of storage after page load.

Reply:
Ah, right. In the multiple tabs scenario, the worst that might happen is that they both get initialised from the same data, but end up wanting to write different data afterwards. I don't think this PR should have any effect on that.

The only thing that this PR tries to change is the time at which the initial data gets read; all of it will be read at the same time, now, rather than on separate occasions, immediately before the corresponding reducer gets initialised.

Reply:
> As far as I understand, the persisted data is only used as a way of showing something quickly, but it quickly gets replaced by fresh data from the API.

💯